### ISSUE TITLE
ParameterBinderFactory/Binders not working as expected

### ISSUE THREAD

**Issue by emanresusername** [scalikejdbc/scalikejdbc#530]
on version `2.4.1`
to keep this DRY, the core code snippets and problem scenario are [here](https://github.com/lloydmeta/enumeratum/issues/37)
i noticed that the [`TypeBinder`'s type parameter is a `+A`](https://github.com/scalikejdbc/scalikejdbc/blob/2.4.1/scalikejdbc-core/src/main/scala/scalikejdbc/TypeBinder.scala#L12) where as the [`ParameterBinderFactory`'s is just `A`](https://github.com/scalikejdbc/scalikejdbc/blob/2.4.1/scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala#L35)
maybe that's related?

**Comment by seratch** [comment]
I'm not sure what you really need on scalikejdbc side yet.

**Comment by emanresusername** [comment]
yeah, so i guess i'm wondering if the signature of `ParameterBinderFactory` was
```scala
ParameterBinderFactory[+A] { ...
```
instead of 
```scala
ParameterBinderFactory[A] { ...
```
if that'd address the issue? 
could contribute a more substantive PR if i'm on the right track with that?

**Comment by seratch** [comment]
@gakuzzzz what do you think?

**Comment by toggm** [comment]
I agree with emanresusername.
I've got the same problem

sealed trait XXX
case object YYY extends XXX

With a parameterbinderfactory defined for XXX. The queries won't allow to out YYY into an eq. I have to write it as YYY.asInstanceof[XXX].

**Comment by gakuzzzz** [comment]
It needs not `ParameterBinderFactory[+A]` but `ParameterBinderFactory[-A]`.

Now, ParameterBinderFactory has `xmap`. the xmap can not define with a contravariant type parameter.

Fundamentally, ParameterBinderFactory's type parameter should be contravariant.
However it breaks compatibility and logging system.

**Comment by gakuzzzz** [comment]
@emanresusername @toggm

I'v tried making the type parameter of ParamaterBinderFactory contravariant.
Unfortunately, it could not do since java.util.Date and java.sql.Tmestamp conflicts.

Now, there is a workaround as follows.

```scala
sealed trait XXX
case object YYY extends XXX

object XXX {
  implicit def xxxPBF[A <: XXX]: ParameterBinderFactory[A] = {
    ParameterBinderFactory(v => (stmt, idx) => stmt.setString(idx, v.toString))
  }
}
```

I think it is better that it remove a type parameter from ParameterBinderWithValue and introduce `contramap` to ParameterBinderFactory. 
like as https://github.com/gakuzzzz/scalikejdbc/commit/748fabed2203129adc21d112f6e6ea284be6cc61

if so, we can define that easily as follows.

```scala
  implicit def xxxPBF[A <: XXX]: ParameterBinderFactory[A] = Binders.string.contramap(_.toString)
``` 

What do you think? @seratch @xuwei-k

**Comment by toggm** [comment]
btw. did you see my issue about the parameterbinderfactory support in the scalikejdbc-async project?
https://github.com/scalikejdbc/scalikejdbc-async/issues/36

**Comment by seratch** [comment]
>What do you think? @seratch @xuwei-k

@gakuzzzz Looks acceptable for me. BTW, it'd be better to discuss the suggestion on another issue or pull request.

@toggm Thanks. We should investigate into it. If you'd helped me on it, it'd be greatly appreciated.

**Comment by unoexperto** [comment]
@gakuzzzz What do you mean by "However it breaks compatibility and logging system." ?

**Comment by seratch** [comment]
This issue should be fixed by #536. If we have something further regarding this topic, please create a new issue for it.
